File: voice-bridge/providers/gemini_live.py
"""Google Gemini Live provider — different protocol from OpenAI."""
from __future__ import annotations

import asyncio
import base64
import json
import logging
import struct
from typing import AsyncIterator

import websockets

from config import ProviderConfig
from providers.base import (
    AudioEvent,
    ErrorEvent,
    ProviderEvent,
    ToolCallEvent,
    TranscriptEvent,
    VoiceProvider,
)

logger = logging.getLogger(__name__)


def _openai_tools_to_gemini(tools: list[dict]) -> list[dict]:
    """Convert OpenAI function tool format to Gemini function declarations."""
    declarations = []
    for tool in tools:
        if tool.get("type") != "function":
            continue
        fn = tool["function"]
        declarations.append({
            "name": fn["name"],
            "description": fn.get("description", ""),
            "parameters": fn.get("parameters", {}),
        })
    return declarations


class GeminiLiveProvider(VoiceProvider):
    def __init__(self, config: ProviderConfig) -> None:
        self.config = config
        self._ws: websockets.ClientConnection | None = None

    async def connect(self, instructions: str, tools: list[dict], vad: bool = False) -> None:
        model = self.config.model
        url = f"{self.config.url}?key={self.config.api_key}"

        self._ws = await websockets.connect(url)

        # Build setup message
        setup: dict = {
            "model": f"models/{model}",
            "generationConfig": {
                "responseModalities": ["AUDIO"],
                "speechConfig": {
                    "voiceConfig": {
                        "prebuiltVoiceConfig": {
                            "voiceName": self.config.voice,
                        }
                    }
                },
            },
            "systemInstruction": {
                "parts": [{"text": instructions}],
            },
        }

        if tools:
            gemini_fns = _openai_tools_to_gemini(tools)
            if gemini_fns:
                setup["tools"] = [{"functionDeclarations": gemini_fns}]

        await self._ws.send(json.dumps({"setup": setup}))

        # Wait for setupComplete
        raw = await self._ws.recv()
        msg = json.loads(raw)
        if "setupComplete" in msg:
            logger.info("Gemini session ready")
        else:
            logger.warning("Unexpected first message: %s", list(msg.keys()))

    async def send_audio(self, audio_b64: str) -> None:
        if not self._ws:
            return
        await self._ws.send(json.dumps({
            "realtime_input": {
                "audio": {
                    "data": audio_b64,
                    "mimeType": "audio/pcm;rate=16000",
                },
            },
        }))

    async def commit_audio(self) -> None:
        # Gemini uses server-side VAD — no client_content commit is possible.
        # But VAD needs silence in the audio stream to detect end-of-turn.
        # When the browser stops recording, audio stops flowing and VAD never
        # triggers. Send ~500ms of silence so VAD sees speech→silence.
        if not self._ws:
            return
        n_samples = int(16000 * 0.1)  # 100ms at 16kHz
        silence = base64.b64encode(struct.pack(f"<{n_samples}h", *([0] * n_samples))).decode()
        for _ in range(5):
            await self._ws.send(json.dumps({
                "realtime_input": {
                    "audio": {
                        "data": silence,
                        "mimeType": "audio/pcm;rate=16000",
                    },
                },
            }))
            await asyncio.sleep(0.1)

    async def receive(self) -> AsyncIterator[ProviderEvent]:
        if not self._ws:
            return
        try:
            async for raw in self._ws:
                msg = json.loads(raw)

                if "serverContent" in msg:
                    content = msg["serverContent"]
                    model_turn = content.get("modelTurn", {})
                    parts = model_turn.get("parts", [])

                    for part in parts:
                        if "inlineData" in part:
                            data = part["inlineData"]
                            if data.get("mimeType", "").startswith("audio/"):
                                yield AudioEvent(audio_b64=data["data"])
                        # Gemini native audio emits text as internal reasoning,
                        # not speech transcript — skip it to avoid leaking
                        # chain-of-thought to the UI.

                    if content.get("turnComplete"):
                        yield TranscriptEvent(text="", role="assistant", final=True)

                elif "toolCall" in msg:
                    tc = msg["toolCall"]
                    for fn_call in tc.get("functionCalls", []):
                        yield ToolCallEvent(
                            call_id=fn_call.get("id", ""),
                            name=fn_call.get("name", ""),
                            arguments=json.dumps(fn_call.get("args", {})),
                        )

                elif "setupComplete" in msg:
                    pass

                elif "interrupted" in msg:
                    logger.debug("Gemini response interrupted")

                else:
                    logger.debug("Unhandled Gemini message: %s", list(msg.keys()))

        except websockets.ConnectionClosed:
            logger.info("Gemini WebSocket closed")

    async def send_tool_result(self, call_id: str, result: str) -> None:
        if not self._ws:
            return
        await self._ws.send(json.dumps({
            "tool_response": {
                "functionResponses": [{
                    "id": call_id,
                    "name": "",  # Gemini doesn't require name in response
                    "response": {"result": result},
                }],
            },
        }))

    async def close(self) -> None:
        if self._ws:
            await self._ws.close()
            self._ws = None

File: dashboard/src/voice/providers/gemini-live.ts
import WebSocket from 'ws'
import type { ProviderConfig } from '../config'
import type { ProviderEvent, ToolDef, VoiceProvider } from './base'

function openaiToolsToGemini(tools: ToolDef[]): Record<string, unknown>[] {
  return tools
    .filter((t) => t.type === 'function')
    .map((t) => ({
      name: t.function.name,
      description: t.function.description || '',
      parameters: t.function.parameters || {},
    }))
}

function generateSilenceB64(): string {
  // 100ms of silence at 16kHz, signed 16-bit LE
  const nSamples = 1600
  const buf = Buffer.alloc(nSamples * 2) // all zeros = silence
  return buf.toString('base64')
}

export class GeminiLiveProvider implements VoiceProvider {
  private ws: WebSocket | null = null
  private config: ProviderConfig

  constructor(config: ProviderConfig) {
    this.config = config
  }

  async connect(instructions: string, tools: ToolDef[], _vad = false): Promise<void> {
    const url = `${this.config.url}?key=${this.config.apiKey}`
    this.ws = new WebSocket(url)

    // Wait for open
    await new Promise<void>((resolve, reject) => {
      this.ws!.once('open', resolve)
      this.ws!.once('error', reject)
    })

    // Build setup message
    const setup: Record<string, unknown> = {
      model: `models/${this.config.model}`,
      generationConfig: {
        responseModalities: ['AUDIO'],
        speechConfig: {
          voiceConfig: {
            prebuiltVoiceConfig: { voiceName: this.config.voice },
          },
        },
      },
      systemInstruction: { parts: [{ text: instructions }] },
    }

    if (tools.length > 0) {
      const geminiFns = openaiToolsToGemini(tools)
      if (geminiFns.length > 0) {
        setup.tools = [{ functionDeclarations: geminiFns }]
      }
    }

    this.ws.send(JSON.stringify({ setup }))

    // Wait for setupComplete
    await new Promise<void>((resolve, reject) => {
      this.ws!.once('message', (raw) => {
        const msg = JSON.parse(raw.toString())
        if ('setupComplete' in msg) {
          resolve()
        } else {
          reject(new Error(`Unexpected first message: ${JSON.stringify(Object.keys(msg))}`))
        }
      })
      this.ws!.once('error', reject)
    })
  }

  async sendAudio(audioB64: string): Promise<void> {
    this.ws?.send(
      JSON.stringify({
        realtime_input: {
          audio: { data: audioB64, mimeType: 'audio/pcm;rate=16000' },
        },
      }),
    )
  }

  async commitAudio(): Promise<void> {
    // Gemini uses server-side VAD only. Send ~500ms of silence so VAD
    // sees speech→silence and triggers end-of-turn.
    if (!this.ws) return
    const silence = generateSilenceB64()
    for (let i = 0; i < 5; i++) {
      this.ws.send(
        JSON.stringify({
          realtime_input: {
            audio: { data: silence, mimeType: 'audio/pcm;rate=16000' },
          },
        }),
      )
      await new Promise((r) => setTimeout(r, 100))
    }
  }

  async *receive(): AsyncGenerator<ProviderEvent> {
    if (!this.ws) return

    for await (const raw of this.iterateMessages()) {
      const msg = JSON.parse(raw)

      if ('serverContent' in msg) {
        const content = msg.serverContent
        const parts = content.modelTurn?.parts || []
        for (const part of parts) {
          if (part.inlineData?.mimeType?.startsWith('audio/')) {
            yield { kind: 'audio', audioB64: part.inlineData.data }
          }
          // Text from Gemini native audio is internal reasoning — skip it
        }
        if (content.turnComplete) {
          yield { kind: 'transcript', text: '', role: 'assistant', final: true }
        }
      } else if ('toolCall' in msg) {
        for (const fnCall of msg.toolCall.functionCalls || []) {
          yield {
            kind: 'tool_call',
            callId: fnCall.id || '',
            name: fnCall.name || '',
            arguments: JSON.stringify(fnCall.args || {}),
          }
        }
      } else if ('interrupted' in msg) {
        // logged, not surfaced
      }
      // setupComplete and others silently ignored
    }
  }

  async sendToolResult(callId: string, result: string): Promise<void> {
    this.ws?.send(
      JSON.stringify({
        tool_response: {
          functionResponses: [
            { id: callId, name: '', response: { result } },
          ],
        },
      }),
    )
  }

  async close(): Promise<void> {
    if (this.ws) {
      this.ws.close()
      this.ws = null
    }
  }

  private async *iterateMessages(): AsyncGenerator<string> {
    const ws = this.ws!
    const queue: string[] = []
    let resolve: (() => void) | null = null
    let done = false

    const onMessage = (data: WebSocket.Data) => {
      queue.push(data.toString())
      resolve?.()
    }
    const onClose = () => {
      done = true
      resolve?.()
    }
    const onError = () => {
      done = true
      resolve?.()
    }

    ws.on('message', onMessage)
    ws.on('close', onClose)
    ws.on('error', onError)

    try {
      while (true) {
        while (queue.length > 0) yield queue.shift()!
        if (done) return
        await new Promise<void>((r) => { resolve = r })
        resolve = null
      }
    } finally {
      ws.off('message', onMessage)
      ws.off('close', onClose)
      ws.off('error', onError)
    }
  }
}
